### PR TITLE
[torch.compile] Add explain as a backend #102053

### LINKED ISSUE #102053: [torch.compile] Add `explain` as a backend
[`torch._dynamo.explain`](https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/_dynamo/eval_frame.py#L562) is a useful API for PT2 users. Users can find the number of graphs and graph breaks using this API, giving them actionable messages. One of the pain points is - its a different API than torch.compile. @Chillee  proposed to add it as a backend, simplifying the UX issue. 

~~~
import torch
import torch._dynamo

def fn3(x):
    x = torch.sin(x)
    torch._dynamo.graph_break()
    x = torch.sin(x)
    return x



def fn2(x):
    x = torch.cos(x)
    x = fn3(x)
    x = torch.cos(x)
    return x


def fn1(x):
    x = torch.tan(x)
    torch._dynamo.graph_break()
    x = fn2(x)
    x = torch.tan(x)
    return x

def fn(x):
    x = torch.sigmoid(x)
    x = fn1(x)
    x = torch.sigmoid(x)
    return x


# outs = torch._dynamo.explain(fn, torch.randn(4))
explanation, out_guards, graphs, ops_per_graph, break_reasons, explanation_verbose = torch._dynamo.explain(fn, torch.randn(4))
print(explanation_verbose)
for graph in graphs:
    print(graph.graph.print_tabular())

~~~    

### Short term task

Introduce a new backend - `explain` such that `torch.compile(fn, backend="explain")` gives morally same output as torch._dynamo.explain. We can add the `explain` backend in this file - [torch/_dynamo/backends/debugging.py](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/backends/debugging.py)

As a start, we can focus on graph breaks and explanation string. We would want the backend to be configurable so that we can expand to other stuff.

### Improving the output

* More details around graph breaks. For example, a graph break deep in the call stack can lead to many graph breaks in the model. The output should explain this part well, and potentially rank the graph breaks.
* Recompiles - And reason of recompile

cc @ezyang @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by laithhas**
Fixes #102053

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @aakhundov

### CODE REVIEW

**Comment by laithhas** [comment]
As discussed in the issue tied to this PR, starting off with just creating the interface and basic graph break/graph/op info. In the future, we can extend this to make it more useful. We can also configure it so that users can specify which information they want to see by passing flags. I can make these changes in upcoming PRs.

Example program: P764281511

New explain backend output: P764280785

Previous torch._dynamo.explain() output: P764280593

File: torch/_dynamo/backends/debugging.py [hunk 1]
```diff
@@ -132,3 +132,64 @@ def non_leaf_compile_error_TESTING_ONLY(gm: torch.fx.GraphModule, example_inputs
...
+
+        torch._dynamo.reset()
+        eb = ExplainWithBackend("inductor")
+        optimized_fn = torch.compile(fn, backend=eb)
+        result = optimized_fn(torch.randn(5))
+        print(eb)
+    """
+
+    def __init__(self, backend):
+        self.backend = backend
```
Comment (voznesenskym) [inline torch/_dynamo/backends/debugging.py:163]: assert that backend is real here, waiting for `lookup_backend` works, but it would be nicer to know at instantiation time.

File: torch/_dynamo/backends/debugging.py [hunk 2]
```diff
@@ -132,3 +132,64 @@ def non_leaf_compile_error_TESTING_ONLY(gm: torch.fx.GraphModule, example_inputs
...
+        from .registry import lookup_backend
+
+        self.graphs.append(gm)
+        ops = []
+        for node in gm.graph.nodes:
+            if node.op == "call_function":
+                ops.append(node.target)
+        self.op_count += len(ops)
+        if gm.compile_subgraph_reason.graph_break:
+            self.break_reasons.append(gm.compile_subgraph_reason)
```
Comment (voznesenskym) [inline torch/_dynamo/backends/debugging.py:178]: I said I would reject any PR that duplicates code from `def explain`, so, I will reject for this here.

Part of doing this the right way is taking the `dynamo_graph_accumulating_compiler` and deduplicating it with this so that explain output is nice and uniform.

File: torch/_dynamo/backends/debugging.py [hunk 3]
```diff
@@ -132,3 +132,64 @@ def non_leaf_compile_error_TESTING_ONLY(gm: torch.fx.GraphModule, example_inputs
...
+
+        return lookup_backend(self.backend)(gm, example_inputs)
+
+    def __str__(self):
```
Comment (voznesenskym) [inline torch/_dynamo/backends/debugging.py:182]: Not bad, but I want to see you utilize the structure @msaroufim added here https://github.com/pytorch/pytorch/pull/102869/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911R558

**Review by voznesenskym — CHANGES_REQUESTED** [review]
Comments inline, back to sender.

Also, add a test.

File: torch/_dynamo/backends/debugging.py (hunk 1 above)
Comment (laithhas) [reply torch/_dynamo/backends/debugging.py]: Sure, I updated it so that `lookup_backend` is called in `__init__`. If there is an issue then the `InvalidBackend` exception will be raised internally by `lookup_backend`

File: torch/_dynamo/backends/debugging.py (hunk 2 above)
Comment (laithhas) [reply torch/_dynamo/backends/debugging.py]: Got it, you are right — had to do some finessing of the parameters but made it work.

File: torch/_dynamo/backends/debugging.py (hunk 3 above)
Comment (laithhas) [reply torch/_dynamo/backends/debugging.py:241]: Updated the structure and made sure it's used in both places (y)

File: torch/_dynamo/eval_frame.py [hunk 4]
```diff
@@ -666,13 +620,13 @@ def guard_export_print(guards):
...
         graph_break_count,
+        break_reasons,
         op_count,
-        out_guards,
-        graphs,
         ops_per_graph,
-        break_reasons,
+        out_guards,
         compile_time,
     )
```
Comment (voznesenskym) [inline torch/_dynamo/eval_frame.py:631]: why the motion here

File: torch/_dynamo/backends/debugging.py [hunk 5]
```diff
@@ -132,3 +135,134 @@ def non_leaf_compile_error_TESTING_ONLY(gm: torch.fx.GraphModule, example_inputs
...
+            for i, guard in enumerate(self.out_guards):
+                output += f"  Guard {i+1}:\n"
+                output += f"    {str(guard)}"
+
+        if self.compile_times is not None:
+            output += f"Compile Times: {self.compile_times}\n"
+        return output
+
+
+def process_graph(
```
Comment (voznesenskym) [inline torch/_dynamo/backends/debugging.py:189]: nit - _explain_graph_detail or something like that

File: torch/_dynamo/backends/debugging.py [hunk 6]
```diff
@@ -132,3 +135,134 @@ def non_leaf_compile_error_TESTING_ONLY(gm: torch.fx.GraphModule, example_inputs
...
+        self.op_count = 0
+        self.break_reasons = []
+
+    def __call__(self, gm: torch.fx.GraphModule, example_inputs):
+        gm, self.graphs, self.op_count, _, self.break_reasons = process_graph(
+            gm, self.graphs, self.op_count, [], self.break_reasons
+        )
+        return self.backend(gm, example_inputs)
+
+    def __str__(self):
```
Comment (voznesenskym) [inline torch/_dynamo/backends/debugging.py:258]: I still don't think we should `__str__` it, leave it as an exercise to the reader to do this if they want, since it just involves calling `str` on output.

Instead, add an api like `def output() -> ExplainOutput`

**Review by voznesenskym — COMMENTED** [review]
Very close, just needs a test. Otherwise looks good.

File: torch/_dynamo/eval_frame.py (hunk 4 above)
Comment (laithhas) [reply torch/_dynamo/eval_frame.py:631]: My OCD kicked in and I updated the order of the ExplainOutput attributes to be grouped together more logically :)

**Review by voznesenskym — APPROVED** [review]

**Comment by laithhas** [comment]
@voznesenskym @anijain2305

Not sure what the merge process is since this is my first time. I have added myself as a contributor to the PyTorch org but am still seeing "The base branch restricts merging to authorized users"

Also are the failing CI jobs noise or legit? I looked into them and they seem to be unrelated to any changes I made.

Thanks for helping me get my first PR done!
